src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseAllocation.kt

```diff
@@ -0,0 +1,125 @@
+package uk.gov.justice.digital.hmpps.nomisprisonerapi.helper.builders
+
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.CourseActivity
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.CourseSchedule
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderBooking
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderProgramProfile
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderProgramStatus
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.ProgramServiceEndReason
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderProgramProfileRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ReferenceCodeRepository
+import java.time.LocalDate
+
+@DslMarker
+annotation class CourseAllocationDslMarker
+
+@TestDataDslMarker
+interface CourseAllocationDsl {
+
+  @CourseAllocationPayBandDslMarker
+  fun payBand(
+    startDate: String = "2022-10-31",
+    endDate: String? = null,
+    payBandCode: String = "5",
+  )
+
+  @CourseAttendanceDslMarker
+  fun courseAttendance(
+    courseSchedule: CourseSchedule,
+    eventId: Long = 0,
+    eventStatusCode: String = "SCH",
+    toInternalLocationId: Long? = -8,
+    outcomeReasonCode: String? = null,
+    paidTransactionId: Long? = null,
+  )
+}
+
+@Component
+class CourseAllocationBuilderRepository(
+  private val offenderProgramProfileRepository: OffenderProgramProfileRepository,
+  private val programStatusRepository: ReferenceCodeRepository<OffenderProgramStatus>,
+  private val programEndReasonRepository: ReferenceCodeRepository<ProgramServiceEndReason>,
+) {
+  fun save(courseAllocation: OffenderProgramProfile) = offenderProgramProfileRepository.save(courseAllocation)
+  fun programStatus(code: String): OffenderProgramStatus = programStatusRepository.findByIdOrNull(OffenderProgramStatus.pk(code))!!
+  fun programEndReason(code: String): ProgramServiceEndReason = programEndReasonRepository.findByIdOrNull(ProgramServiceEndReason.pk(code))!!
+}
+
+@Component
+class CourseAllocationBuilderFactory(
+  private val repository: CourseAllocationBuilderRepository,
+  private val payBandBuilderFactory: CourseAllocationPayBandBuilderFactory,
+  private val courseAttendanceBuilderFactory: CourseAttendanceBuilderFactory,
+) {
```

Comment:
This is an example of a parent entity that has its child builder factories injected by Spring along with its own dedicated repository.